R/data_Validate.R

```diff
@@ -0,0 +1,93 @@
+# We might want to do this entirely via workflows, but I feel like it's good to
+# check in the app itself since users don't HAVE to use workflows.
+
+validate_df <- function(df,
+                        chrRequiredColumns = character(),
+                        arg = rlang::caller_arg(df),
+                        call = rlang::caller_env()) {
+  df <- validate_is_df(df, arg = arg, call = call)
+  validate_has_columns(
+    df,
+    chrRequiredColumns = chrRequiredColumns,
+    arg = arg,
+    call = call
+  )
+}
+
+validate_is_df <- function(df,
+                           arg = rlang::caller_arg(df),
+                           call = rlang::caller_env()) {
+  if (is.data.frame(df)) {
+    return(df)
+  }
+  cli::cli_abort(
+    c(
+      "{.arg {arg}} must be a data.frame.",
+      x = "{.arg {arg}} is {.obj_type_friendly {df}}."
+    ),
+    class = "gsm.app-error-invalid_input",
+    call = call
+  )
+}
+
+validate_has_columns <- function(df,
+                                 chrRequiredColumns = character(),
+                                 arg = rlang::caller_arg(df),
+                                 call = rlang::caller_env()) {
+  missing_cols <- setdiff(chrRequiredColumns, colnames(df))
+  if (length(missing_cols)) {
+    cli::cli_abort(
+      c(
+        "{.arg {arg}} must have all required columns.",
+        x = "Missing columns: {.field {missing_cols}}."
+      ),
+      class = "gsm.app-error-invalid_input"
+    )
+  }
+  return(df)
+}
+
+## df-specific Validation Functions ----
+
+validate_dfResults <- function(dfResults) {
```

Comment:
for these functions, i agree that it makes sense to have fixed required columns at the moment, but is there a future state where adding additional optional modules/tabs to the app would require a different set of required columns? in which case, having a `chrRequiredColumns` with defaults set to the current set of columns. Not anything necessary at the moment- and maybe too much future-proofing, but just thought i'd raise the point while i'm reviewing this.

Reply:
I think this app will have requirements, even if those requirements might change; we can explicitly change the list inside the function when they do. Once we have some of this information in workflows, though, it MIGHT make sense to check there so everything is logged in one place. Mostly I want to make sure we aren't just counting on people to use our preferred method of deploying things, and then crashing downstream when they don't use that method and something is weird.